xdsl/transforms/printf_to_putchar.py

```diff
@@ -0,0 +1,243 @@
+        * It does not require a printf implementation, only putchar.
+    """
+
+    # get_number_of_digits implementation
+    def get_number_of_digits(absolute_value: Operation) -> OpResult:
+        """
+        Return the result of an MLIR scf.while op that gets the number of
+        digits of a positive integer. Roughly goes like this in
+        python pseudocode (e.g. integer is 420, routine should return 3):
+        ```
+        integer = 420
+        digits = 0
+        while(integer != 0):
+            integer //= 10
+            digits += 1
+        return digits
+
+        ```
+        """
+        before_block = Block(arg_types=(i32, i32))
+        after_block = Block(arg_types=(i32, i32))
+        digit_init = arith.Constant.from_int_and_width(0, i32)
+        one = arith.Constant.from_int_and_width(1, i32)
+        while_loop = scf.While(
+            [absolute_value, digit_init],
+            [i32, i32],
+            [before_block],
+            [after_block],
+        )
+        with ImplicitBuilder(before_block) as (running_integer, digits):
+            # Stop when you reach zero
+            is_zero = arith.Cmpi(running_integer, zero, "ne")
+            scf.Condition.get(is_zero, running_integer, digits)
+        with ImplicitBuilder(after_block) as (running_integer, previous_digits):
+            ten = arith.Constant.from_int_and_width(10, 32)
+            new_integer = arith.DivUI(running_integer, ten)
+            digits = arith.Addi(previous_digits, one)
+            scf.Yield.get(new_integer, digits)
+
+        return while_loop.results[1]
+
+    # print_minus_if_negative implementation
+    def print_minus_if_negative(integer: SSAValue) -> scf.If:
+        """
+        This function returns an MLIR scf.if op that prints a minus sign
+        if the integer value is negative and returns the absolute value
+        of the integer in either case, in python code:
+        ```
+        if(integer > 0):
+            print("-")
+            return (0 - integer)
+        else:
+            return integer
+        ```
+        """
+        is_negative_block = Block()
+        is_positive_block = Block()
+        is_negative = arith.Cmpi(integer, zero, "slt")
+
+        # Either way get the absolute value of the number
+        absolute_value = scf.If.get(
+            is_negative, [i32], [is_negative_block], [is_positive_block]
+        )
+        # If the value is negative, print the minus sign, return abs value
+        with ImplicitBuilder(is_negative_block):
+            PrintCharOp.from_constant_char("-")
+            negative_input = arith.Subi(zero, integer)
+            scf.Yield.get(negative_input)
+        # If the value is positive, just return value itself as abs value
+        with ImplicitBuilder(is_positive_block):
+            scf.Yield.get(integer)
+
+        return absolute_value
+
+    def print_digits(digits: SSAValue, absolute_value: scf.If) -> scf.For:
```

Comment:
It looks weird that `absolute_value` has type `scf.If`.

Reply:
Can you please pass an `SSAValue` here instead?

Reply:
Thanks for the remark @kingiler and the suggestion @superlopuh